src/soroban/assembled_transaction.ts

```diff
@@ -0,0 +1,717 @@
+    this.server = new SorobanRpc.Server(this.options.rpcUrl, {
+      allowHttp: this.options.rpcUrl.startsWith("http://"),
+    });
+  }
+
+  static async fromSimulation<T>(
+    options: AssembledTransactionOptions<T>
+  ): Promise<AssembledTransaction<T>> {
+    const tx = new AssembledTransaction(options);
+    const contract = new Contract(options.contractId);
+
+    tx.raw = new TransactionBuilder(await tx.getAccount(), {
+      fee: options.fee?.toString(10) ?? BASE_FEE,
+      networkPassphrase: options.networkPassphrase,
+    })
+      .addOperation(contract.call(options.method, ...(options.args ?? [])))
+      .setTimeout(TimeoutInfinite)
+      .build();
+
+    return await tx.simulate();
+  }
+
+  simulate = async (): Promise<this> => {
+    if (!this.raw) throw new Error("Transaction has not yet been assembled");
+    this.simulation = await this.server.simulateTransaction(this.raw);
+
+    if (SorobanRpc.Api.isSimulationSuccess(this.simulation)) {
+      this.raw = SorobanRpc.assembleTransaction(
+        this.raw,
+        this.simulation
+      ).build();
+    }
+
+    return this;
+  };
+
+  get simulationData(): {
+    result: SorobanRpc.Api.SimulateHostFunctionResult;
+    transactionData: xdr.SorobanTransactionData;
+  } {
+    if (this.simulationResult && this.simulationTransactionData) {
+      return {
+        result: this.simulationResult,
+        transactionData: this.simulationTransactionData,
+      };
+    }
+    // else, we know we just did the simulation on this machine
+    const simulation = this.simulation!;
+    if (SorobanRpc.Api.isSimulationError(simulation)) {
+      throw new Error(`Transaction simulation failed: "${simulation.error}"`);
+    }
+
+    if (SorobanRpc.Api.isSimulationRestore(simulation)) {
+      throw new AssembledTransaction.ExpiredStateError(
+        `You need to restore some contract state before you can invoke this method. ${JSON.stringify(
+          simulation,
+          null,
+          2
+        )}`
+      );
+    }
+
+    if (!simulation.result) {
+      throw new Error(
+        `Expected an invocation simulation, but got no 'result' field. Simulation: ${JSON.stringify(
+          simulation,
+          null,
+          2
+        )}`
+      );
+    }
+
+    // add to object for serialization & deserialization
+    this.simulationResult = simulation.result;
+    this.simulationTransactionData = simulation.transactionData.build();
+
+    return {
+      result: this.simulationResult,
+      transactionData: this.simulationTransactionData!,
+    };
+  }
+
+  get result(): T {
+    try {
+      return this.options.parseResultXdr(this.simulationData.result.retval);
+    } catch (e) {
+      if (!implementsToString(e)) throw e;
+      let err = this.parseError(e.toString());
+      if (err) return err as T;
+      throw e;
+    }
+  }
+
+  parseError(errorMessage: string): Result<never, ErrorMessage> | undefined {
```

Comment:
I would rather return `errorMessage` than `undefined` in the mismatched cases.

Reply:
Error the error should align with the contract's errors. So would you prefer it throws instead? Otherwise not sure which contract error to give it.